### PR TITLE
Use get_fresh_aux_symbol to construct dynamic objects

### PR DESCRIPTION
**Pull request by tautschnig**
This avoids use of a global variable and reuses central infrastructure instead
of repeated local work. Note that this changes the names of dynamic objects as
the suffix now includes a $ character, and need not have any suffix, as well as
moving the counter out of the middle of C++/Java dynamic_X_array objects.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 3dccbb324915003e2b402059d157123b90bc0383).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99183099

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: a6180e68ee227cd12e7647b9f60d47bf2cd4e6e7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99244949

**Review by peterschrammel — COMMENTED** [review]
Thanks, @tautschnig, for picking this up. We have to verify whether this breaks downstream consumption of traces.

**Comment by tautschnig** [comment]
Marking do-not-merge until approved by users. @peterschrammel can you please co-ordinate this?

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: a6ef0d571795900dc5f29b0deaf728b4cce9eefc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99430635
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 6fa9d82dcea8345e951e642f538c82e0e42ffe0c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99433156

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: bde95d04d6b3202a3264ebd4be8991650567ce8f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99678277
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 375f0520f8841b61cfc2d2cd064317cc31e4f8ee).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99704183

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 2e2ffd4d74bbc5644ae07f3403394a4744c0c50c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100068548

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: d39c8d4a52ef03c1d53fc5e40701556c7ed0dd39).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100079456

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 96c47526df4595a7e2b57e34ac56f9d9c7cb12ad).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100652670

**Review by peterschrammel — APPROVED** [review]
@thk123, @allredj, can this be merged?

**Comment by tautschnig** [comment]
This is rebased, and now requires an additional review by @kroening. @allredj @thk123 If you, or members of your teams could check whether this still breaks TG that would be helpful. @svorenova had started looking into this a while ago.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 65442e04849eea4af7551acaf5f55a20d2a1412d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/103512122

**Comment by majakusber** [comment]
@tautschnig I just checked, there is still at least one regression test failing (the same one I observed last time).

**Comment by tautschnig** [comment]
Thanks @svorenova! My suspicion, but obviously I don't have insight into that, is that there is some assumption that strings like `dynamic_1_array` are generated, which is no longer true as of this PR.

**Comment by majakusber** [comment]
@tautschnig Yes, I was looking for that sort of issue when I was originally investigating the failing test but it wasn't clear what the problem was coming from. I'll take another look and let you know.

**Comment by tautschnig** [comment]
I'm probably very naive about this, but I thought the only externally visible change was
```
dynamic_1_array -> dynamic_array (or dynamic_array$1 for the second, third, etc)
dynamic_1_value -> dynamic_value (or dynamic_value$1 for the second, third, etc)
```
But maybe I messed something else up in make these changes here.

**Comment by allredj** [comment]
@tautschnig is this still something you want to do?

**Comment by tautschnig** [comment]
> @tautschnig is this still something you want to do?

@allredj Yes, I'd still like to get this merged at some point. Unfortunately I'm getting very little engineering time at the moment, so I'm not sure whether I can promise a rebase over the next days. But the TG debugging probably still needs to happen. I guess the next step is on me to get this rebased first?

**Comment by allredj** [comment]
@tautschnig yes a rebase will be necessary for us to look at it further. But no rush obviously! :-)

**Comment by allredj** [comment]
@tautschnig Here comes my biannual ping about this PR. It seems to me that symex has changed enough in the past year that it will be more efficient to redo this rather than try to rebase this PR. Would you agree?
Do you want to keep it open?

Btw, I wish you a great Christmas holiday!

**Comment by TGWDB** [comment]
@tautschnig This PR could be nice to rebase and merge, but maybe is too old and deprecated now? Any chance of an update on the status?

**Comment by tautschnig** [comment]
> @tautschnig This PR could be nice to rebase and merge, but maybe is too old and deprecated now? Any chance of an update on the status?

Finally done. Does this still need or want some Diffblue-internal review? If so, @TGWDB can you perhaps co-ordinate for that to happen?

**Review by TGWDB — APPROVED** [review]

**Pull request merged**
